node/core/bitfield-signing/src/lib.rs

```diff
@@ -181,85 +169,55 @@ async fn get_core_availability(
+/// delegates to the v1 runtime API
 async fn get_availability_cores(relay_parent: Hash, sender: &mut mpsc::Sender<FromJob>) -> Result<Vec<CoreState>, Error> {
-	use FromJob::RuntimeApi;
-	use messages::{
-		RuntimeApiMessage::Request,
-		RuntimeApiRequest::AvailabilityCores,
-	};
-
 	let (tx, rx) = oneshot::channel();
-	sender.send(RuntimeApi(Request(relay_parent, AvailabilityCores(tx)))).await?;
+	sender.send(RuntimeApiMessage::Request(relay_parent, RuntimeApiRequest::AvailabilityCores(tx)).into()).await?;
 	match rx.await {
 		Ok(Ok(out)) => Ok(out),
 		Ok(Err(runtime_err)) => Err(runtime_err.into()),
 		Err(err) => Err(err.into())
 	}
 }
 
-// - get the list of core states from the runtime
-// - for each core, concurrently determine chunk availability (see `get_core_availability`)
-// - return the bitfield if there were no errors at any point in this process
-//   (otherwise, it's prone to false negatives)
+/// - get the list of core states from the runtime
+/// - for each core, concurrently determine chunk availability (see `get_core_availability`)
+/// - return the bitfield if there were no errors at any point in this process
+///   (otherwise, it's prone to false negatives)
 async fn construct_availability_bitfield(
 	relay_parent: Hash,
 	validator_idx: ValidatorIndex,
 	sender: &mut mpsc::Sender<FromJob>,
 ) -> Result<AvailabilityBitfield, Error> {
-	use futures::lock::Mutex;
-
 	// get the set of availability cores from the runtime
 	let availability_cores = get_availability_cores(relay_parent, sender).await?;
 
-	// we now need sender to be immutable so we can copy the reference to multiple concurrent closures
-	let sender = &*sender;
-
-	// prepare outputs
-	let out = Mutex::new(bitvec!(bitvec::order::Lsb0, u8; 0; availability_cores.len()));
-	// in principle, we know that we never want concurrent access to the _same_ bit within the vec;
-	// we could `let out_ref = out.as_mut_ptr();` here instead, and manually assign bits, avoiding
-	// any need to ever wait to lock this mutex.
-	// in practice, it's safer to just use the mutex, and speed optimizations should wait until
-	// benchmarking proves that they are necessary.
-	let out_ref = &out;
-	let errs = Mutex::new(Vec::new());
-	let errs_ref = &errs;
-
-	// Handle each (idx, core) pair concurrently
-	//
-	// In principle, this work is all concurrent, not parallel. In practice, we can't guarantee it, which is why
-	// we need the mutexes and explicit references above.
-	stream::iter(availability_cores.into_iter().enumerate())
-		.for_each_concurrent(None, |(idx, core)| async move {
-			let availability = match get_core_availability(relay_parent, core, validator_idx, sender).await {
-				Ok(availability) => availability,
-				Err(err) => {
-					errs_ref.lock().await.push(err);
-					return;
-				}
-			};
-			out_ref.lock().await.set(idx, availability);
-		})
-		.await;
+	// Wrap the sender in a Mutex to share it between the futures.
+	let sender = Mutex::new(sender);
```

Comment:
Why not `clone()`?

Reply:
That would increase the capacity by 1 for each clone.

Reply:
Exactly, this would mean that you can always send at least one item.

Reply:
Could one of you guys please send a follow-up to leave a comment for this? I think it is important.